packages/react/src/components/ContextMenu/ContextMenu.js

```diff
@@ -0,0 +1,244 @@
+    );
+
+    return parentNode?.firstChild || null;
+  }
+
+  function handleKeyDown(event) {
+    event.stopPropagation();
+
+    if (match(event, keys.Escape)) {
+      onClose();
+    }
+
+    const currentNode = event.target.parentNode;
+    let nodeToFocus;
+
+    if (match(event, keys.ArrowUp)) {
+      nodeToFocus = getNextNode(currentNode, -1);
+    } else if (match(event, keys.ArrowDown)) {
+      nodeToFocus = getNextNode(currentNode, 1);
+    } else if (match(event, keys.ArrowRight)) {
+      nodeToFocus = getFirstSubNode(currentNode);
+    } else if (match(event, keys.ArrowLeft)) {
+      nodeToFocus = getParentNode(currentNode);
+    }
+
+    focusNode(nodeToFocus);
+
+    if (rest.onKeyDown) {
+      rest.onKeyDown(event);
+    }
+  }
+
+  function handleClick() {
+    onClose();
+  }
+
+  function handleClickOutside() {
+    onClose();
+  }
+
+  function willFit() {
+    if (rootRef?.current?.element) {
+      const bodyWidth = document.body.clientWidth;
+
+      const reverseMap = [...Array(level)].reduce(
+        (acc) => {
+          const endX = acc.lastX + contextMenuWidth * acc.direction;
+          const fits =
+            acc.direction === 1 ? endX < bodyWidth : endX > contextMenuWidth;
+
+          const newDirection = fits ? acc.direction : acc.direction * -1;
+          const newLastX = fits
+            ? endX
+            : acc.lastX + contextMenuWidth * newDirection;
+
+          return {
+            direction: newDirection,
+            lastX: newLastX,
+            map: [...acc.map, newDirection === 1],
+          };
+        },
+        { direction: 1, lastX: x, map: [] }
+      );
+
+      return reverseMap.map[level - 1];
+    }
+
+    return true;
+  }
+
+  useEffect(() => {
+    const topLevelNodes = getValidNodes(rootRef?.current?.element);
+
+    if (topLevelNodes && topLevelNodes.length > 0) {
+      focusNode(topLevelNodes[0].firstChild);
+    }
+
+    setShouldReverse(!willFit());
+  }, [open, x, y]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  const someNodesHaveIcons = React.Children.toArray(children).some(
+    (node) =>
+      node.props.renderIcon ||
+      node.type === ContextMenuSelectableOption ||
+      node.type === ContextMenuRadioGroup
+  );
+
+  const options = React.Children.map(children, (node) => {
+    if (React.isValidElement(node)) {
+      return React.cloneElement(node, {
+        indented: someNodesHaveIcons,
+        level: level,
+        menuX: x,
+      });
+    }
+  });
+
+  const classes = classnames(`${prefix}--context-menu`, {
+    [`${prefix}--context-menu--open`]: open,
+    [`${prefix}--context-menu--reverse`]: shouldReverse,
+  });
+
+  return (
+    <ClickListener onClickOutside={handleClickOutside} ref={rootRef}>
+      <ul
+        className={classes}
+        onKeyDown={handleKeyDown}
+        onClick={handleClick}
+        data-level={level}
+        style={isRootMenu ? { left: `${x}px`, top: `${y}px` } : null}
```

Comment:
these positioning styles are always taking the `x` and `y` prop values, we probably want to save new `x` and `y` values after making calculations in `willFit` so that the menu reverses correctly

![image](https://user-images.githubusercontent.com/8265238/101688776-2183e580-3a32-11eb-97fb-7f92739f0cbe.png)

![image](https://user-images.githubusercontent.com/8265238/101688801-28125d00-3a32-11eb-8fbb-85886c81ee4a.png)